src/ape/api/address.py

```diff
@@ -0,0 +1,64 @@
+from typing import Optional, Type
+
+from .base import abstractdataclass, abstractmethod
+from .providers import ProviderAPI, ReceiptAPI, TransactionAPI
+
+
+@abstractdataclass
+class AddressAPI:
+    _provider: Optional[ProviderAPI] = None
+
+    @property
+    def provider(self) -> ProviderAPI:
+        if not self._provider:
+            raise Exception("Wired incorrectly")
+
+        return self._provider
+
+    @property
+    def _receipt_class(self) -> Type[ReceiptAPI]:
+        return self.provider.network.ecosystem.receipt_class
+
+    @property
+    def _transaction_class(self) -> Type[TransactionAPI]:
+        return self.provider.network.ecosystem.transaction_class
+
+    @property
+    @abstractmethod
+    def address(self) -> str:
+        ...
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__name__} {self.address}>"
+
+    def __str__(self) -> str:
+        return self.address
+
+    @property
+    def nonce(self) -> int:
+        return self.provider.get_nonce(self.address)
+
+    @property
+    def balance(self) -> int:
+        return self.provider.get_balance(self.address)
+
+    @property
+    def code(self) -> bytes:
+        # TODO: Explore caching this (based on `self.provider.network` and examining code)
```

Comment:
Could you use `@lru_cache`?
https://www.geeksforgeeks.org/python-functools-lru_cache/

Reply:
Caching code is a more complex thing, it basically depends on which network you're own and how it was deployed. I can tell you more about how it works, but a simple lru cache wouldn't really work here.